Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,247 @@
+}
+
+bool DropDownList::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == parent()) {
+    if(event->type() == QEvent::KeyPress) {
+      auto e = static_cast<QKeyEvent*>(event);
+      if(e->key() == Qt::Key_Up) {
+        if(isVisible()) {
+          focus_previous();
+          return true;
+        }
+      } else if(e->key() == Qt::Key_Down) {
+        if(isVisible()) {
+          focus_next();
+          return true;
+        }
+      } else if(e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return) {
+        if(isVisible() && m_highlight_index) {
+          on_item_selected(get_widget(*m_highlight_index)->get_value(),
+            *m_highlight_index);
+          return true;
+        }
+      }
+    }
+  }
+  return DropDownWindow::eventFilter(watched, event);
+}
+
+void DropDownList::hideEvent(QHideEvent* event) {
+  m_scroll_area->verticalScrollBar()->setValue(0);
+  if(m_highlight_index) {
+    auto item = m_layout->itemAt(*m_highlight_index);
+    if(item != nullptr) {
+      static_cast<DropDownItem*>(item->widget())->reset_highlight();
+    }
+    m_highlight_index = boost::none;
+  }
+}
+
+void DropDownList::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_Up) {
+    if(isVisible()) {
+      focus_previous();
+    }
+  } else if(event->key() == Qt::Key_Down) {
+    if(isVisible()) {
+      focus_next();
+    }
+  } else if(event->key() == Qt::Key_Escape) {
+    hide();
+  } else if(event->key() == Qt::Key_Enter) {
+    if(m_highlight_index) {
+      on_item_selected(get_widget(*m_highlight_index)->get_value(),
+        *m_highlight_index);
+    }
+  }
+  DropDownWindow::keyPressEvent(event);
+}
+
+connection DropDownList::connect_activated_signal(
+    const ActivatedSignal::slot_type& slot) const {
+  return m_activated_signal.connect(slot);
+}
+
+connection DropDownList::connect_highlighted_signal(
+    const HighlightedSignal::slot_type& slot) const {
+  return m_highlighted_signal.connect(slot);
+}
+
+connection DropDownList::connect_index_selected_signal(
+    const IndexSelectedSignal::slot_type& slot) const {
+  return m_index_selected_signal.connect(slot);
+}
+
+connection DropDownList::connect_value_selected_signal(
+    const ValueSelectedSignal::slot_type& slot) const {
+  return m_value_selected_signal.connect(slot);
+}
+
+QVariant DropDownList::get_value(int index) {
+  if(m_layout->count() - 1 < index) {
+    return QVariant();
+  }
+  return get_widget(index)->get_value();
+}
+
+void DropDownList::insert_item(DropDownItem* item) {
+  m_item_selected_connections.push_back(std::move(
+    item->connect_selected_signal(
+      [=, index = m_layout->count()] (auto value) {
+        on_item_selected(std::move(value), index);
+      })));
+  m_layout->insertWidget(m_layout->count(), item);
+  update_height();
+}
+
+int DropDownList::item_count() const {
+  return m_layout->count();
+}
+
+void DropDownList::remove_item(int index) {
+  if(index > m_layout->count() - 1 || index < 0) {
```

Comment:
This is probably another place where you should pass an unsigned integer as an argument unless, of course, negative indexes are logically possible.